common/network-common/src/main/java/org/apache/spark/network/server/TransportRequestHandler.java

```diff
@@ -213,7 +213,15 @@ public void onData(String streamId, ByteBuffer buf) throws IOException {
         public void onComplete(String streamId) throws IOException {
            try {
              streamHandler.onComplete(streamId);
-             callback.onSuccess(ByteBuffer.allocate(0));
+             callback.onSuccess(streamHandler.getCompletionResponse());
+           } catch (BlockPushNonFatalFailure ex) {
```

Comment:
Shall we catch `BlockPushNonFatalFailure` inside the `onComplete`? We can have a response filed inside `streamHandler` to store the response value. In this way,  `getCompletionResponse` can handle the success and error cases consistently.

Reply:
We still need to invoke `streamHandler#onFailure` in case we get a non-success return code, which handles logging and other things.
Catching the `BlockPushNonFatalFailure` inside `onComplete` makes it hard to tell apart a success and an error case.
I feel that returning a `ByteBuffer` in `StreamCallback#getCompletionResponse` is generic enough since `StreamCallback` is used for multiple purposes.
Returning a `BlockPushNonFatalFailure` in `StreamCallback` feels too specific for only 1 use case of this interface.

Reply:
The current `getCompletionResponse` only returns the `SUCCESS` byte buffer, right?

I didn't mean to return `BlockPushNonFatalFailure` in the `StreamCallback. getCompletionResponse`. What I expect is to allow return `SUCCESS`, `TOO_LATE_BLOCK_PUSH`, etc, byte buffers in `getCompletionResponse`.

Reply:
What I meant is that by throwing the exception, it's much easier to tell apart the success case from the failure case.
It's not straightforward to tell the content of the returned ByteBuffer without decoding it.
Even for decoding, it's not convenient to decode an rpc message inside network-common module.